nomad/eval_endpoint.go

```diff
@@ -234,10 +234,9 @@ func (e *Eval) Ack(args *structs.EvalAckRequest,
 		return err
 	}
 
-	// It's not necessary to cancel evals before Ack returns, but it's done here
-	// to commit canceled evals as close to the Ack'd eval being committed as
-	// possible.
-	return cancelCancelableEvals(e.srv)
+	// Wake up the eval cancelation reaper
+	e.srv.reapCancelableEvalsCh <- struct{}{}
```

Comment:
This could still block if a raft apply is already in progress. I think to asynchronously trigger another loop you want to use a buffered chan:

```suggestion
	// Wake up the eval cancelation reaper
	select {
		case e.srv.reapCancelableEvalsCh <- struct{}{}:
		default:
	}
```

and make the chan buffered of size 1: `make(struct{}, 1)`

This ensures `Eval.Ack` never blocks but that the cancellation loop always runs at least once after an Eval.Ack (either due to the Ack causing a send, or because a prior Ack's sent is already buffered and waiting to be recv'd).

Reply:
Your implementation isn't wrong though, just not always asynchronous.

Reply:
The channel is already buffered, with the thinking that we wanted to wait until the buffer was empty before continuing. But you're right we could avoid blocking entirely in that case b/c we already know the buffer is full. I merged right before your comment unfortunately, but I can fix that up.

Reply:
Take 2 (3?) is here: https://github.com/hashicorp/nomad/pull/15298